### ISSUE TITLE
Make output report customizable

### ISSUE THREAD

**Issue by mrazauskas** [wellwelwel/poku#629]
In [TSTyche repo](https://github.com/tstyche/tstyche) I have 98 test files. If I run them locally in VS Code terminal, the output rendered by first test files cannot be reached any more. If there are any errors or failures, because the output is outside of scroll back buffer of the terminal.

The only way to see the output of a failing test, is to rerun a particular file.

Perhaps in a multi file run it would be better to emit only errors and file names. (Instead of printing out the whole test tree). Like this:

```shell
  ✔ tests/config-install.test.js › 9020.904792ms
  ✔ tests/config-target.test.js › 14780.683209ms
  ✔ tests/config-typescriptPath.test.js › 2426.265125ms
  ✔ tests/config-update.test.js › 2459.254291ms
  ✔ tests/feature-store.test.js › 16204.369208ms
  ✔ tests/feature-typescript-versions.test.js › 39605.810250ms
  ✔ tests/feature-watch.test.js › 70000.974792ms
  ✔ tests/validation-store.test.js › 5521.964250ms
  ✔ tests/validation-target.test.js › 5053.783042ms
  ✔ tests/validation-update.test.js › 231.104250ms
  ✔ tests/validation-watch.test.js › 197.263041ms
```

If a test file is passing, it isn’t important to see its structure. Keeping focus on errors and failures is most important.

(I would love to have dot reporter, of course. But that’s another story.)

**Comment by mrazauskas** [comment]
To clarify more, output for a single file looks right. Very useful to see all the details. I try to point out that it becomes overwhelming in multi file runs.

**Comment by wellwelwel** [comment]
> To clarify more, output for a single file looks right. Very useful to see all the details. I try to point out that it becomes overwhelming in multi file runs.

Do you think we could impose a limit value to decide whether or not to display the successes?

**Comment by mrazauskas** [comment]
Hard to say how verbose a single file can be. So probably keep it verbose for a single file run and make it silent for a multi file run. Optionally, `--verbose` option could be added to enable / disable verbosity.

What you think?

**Comment by wellwelwel** [comment]
A `--reporter` option would be great. @mrazauskas, I've adapted the PR title 🙋🏻‍♂️

**Comment by mrazauskas** [comment]
> A `--reporter` option would be great. @mrazauskas, I've adapted the issue title 🙋🏻‍♂️

By the way, that would be good solution. `--reporter dot` is all what I need ;D

**Comment by mrazauskas** [comment]
Just an idea, the `--reporter` option could allow adjusting verbosity of the reporter. For instance, `--quiet` can be replaced with `--reporter=quiet`.

I was running tests today and thinking that information like “Running tests in parallel” or “Fail fast is enabled” are useful for debugging. Something like `--log=debug`, `--log=info`, `--log=error`, `--log=quiet` could make sense.

**Comment by wellwelwel** [comment]
> `--quiet` can be replaced with `--reporter=quiet`

About `--debug` and `--quiet`, I like the idea of having them as a direct flag due to familiar commands in _CLI_ apps. The idea is that `--quiet` overrides any debug or reporter options.

For example:

```json
{
  "test": "poku --reporter=dot"
}
```

```sh
npm run test -- --quiet
```

---

In addiction, I wondered how other test runners deal with multiple `reporters`, I intend to create this functionality in a way that makes it easy to implement new alternatives 🙋🏻‍♂️

**Comment by mrazauskas** [comment]
> I wondered how other test runners deal with multiple `reporters`

Jest loops through the list of reporters. A reporter must be an object with a set of callbacks like `onRunComplete()` (all are optional). There are several callbacks and they are call with various data during the test run. The object passed to `onRunComplete()` includes data of the whole test run.

In `node:test` custom reporters are based on a stream of events. That’s lovely pattern!

Inspired by that, I adopted an event emitter for TSTyche. Not only reporters are handle through events, but also the exit code or `failFast` logic (it handles a cancelation token passed to the runner).

**Comment by wellwelwel** [comment]
### Reporters' Ideas:

- `mini` (probably the default)
- `full`
- `dot` (inspired by [**Mocha**](https://github.com/mochajs/mocha))

> The names may not be exactly that.

> [!IMPORTANT]
> 
> Both `--debug` and `--quiet` aren't reporters and should be complementary to all reporters.

---

### Assert

Based on **Deno**'s output, filter and highlight different lines when there is more than one line _(regardless of whether it's a string, objects, etc.)_.

**Comment by mrazauskas** [comment]
The default reporter could be named `poku`. This came to my mind as a joke, but why not? ;D

**Comment by wellwelwel** [comment]
Hey, @mrazauskas 🙋🏻‍♂️

Thank you for the time you've dedicated to **Poku**. All the ideas and reports you've brought have really helped **Poku** evolve.

To not sound a spam in multiple notifications, it took longer than I expected, but finally all the issues you opened (including the confusing way of using `--only`, have been/are being solved in the `RC` version). Including a massive (exponential) performance gap in concurrent tests that you indirectly reported in [**TSTyche**](https://github.com/tstyche/tstyche) 🤝
